[Site Isolation] Make permissions policy work when site isolation is enabled

szewai opened the pull request:
#### 1f8bdc8eab5985a9e2a250e26ad9be088a55dafe
<pre>
[Site Isolation] Make permissions policy work when site isolation is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=276203">https://bugs.webkit.org/show_bug.cgi?id=276203</a>
<a href="https://rdar.apple.com/131080252">rdar://131080252</a>

Reviewed by Youenn Fablet.

When site isolation is enabled, subframe can be in a different process than its parent. However, for document of the
subframe to compute permissions policy, it needs policy information of its owner element (parent frame). To make
permissions policy work with site isolation, we include owner&apos;s policy in navigation data, which is sent from web
process to UI process on deciding policy for navigation. UI process can then forward the owner&apos;s policy to correct frame
process by including it in LoadParameters.

Tests: http/tests/site-isolation/permissions-policy.html
       http/tests/site-isolation/permissions-policy-nested.html

* LayoutTests/http/tests/site-isolation/permissions-policy-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/permissions-policy-nested-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/permissions-policy-nested.html: Added.
* LayoutTests/http/tests/site-isolation/permissions-policy.html: Added.
* LayoutTests/http/tests/site-isolation/resources/permissions-policy-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/resources/permissions-policy-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/permissions-policy-nested-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/simple.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ownerPermissionsPolicy const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::attributeChanged):
(WebCore::HTMLIFrameElement::permissionsPolicyDirective const): Deleted.
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::PermissionsPolicy::computeInheritedPolicyValueInContainer const):
(WebCore::PermissionsPolicy::PermissionsPolicy):
* Source/WebCore/html/PermissionsPolicy.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::setOwnerPermissionsPolicy):
(WebCore::Frame::ownerPermissionsPolicy const):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::ownerPermissionsPolicy const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJSHistoryAPI):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/280942@main">https://commits.webkit.org/280942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59ebe77d4f044974ac507862a8c42a2c56640754

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58078 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10557 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61703 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8523 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60206 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45042 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8712 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47054 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6069 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60108 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35074 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50205 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27888 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31837 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7502 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7527 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53782 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7770 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63404 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7829 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54300 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50218 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54427 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12845 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1704 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33235 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34321 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35405 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34066 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/html/PermissionsPolicy.h:
```diff
@@ -37,11 +37,24 @@ class Document;
+        , documentOrigin(documentOrigin)
+        , containerPolicyValue(WTFMove(containerPolicyValue)) { }
+
+    uint32_t documentInheritedPolicyValue;
```
These should be initialized

youennf replied:
Yes, and it is not really clear what this actually is. If it is a BitSet, why not using a BitSet, or a more precise type?

szewai replied:
The refactoring is done in [280791@main](https://commits.webkit.org/280791@main) and the comments are addressed there.

youennf commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10715,8 +10715,15 @@ PermissionsPolicy Document::permissionsPolicy() const
     // because Document may not be set on Frame yet, and it would affect the computation
     // of PermissionsPolicy.
-    if (!m_permissionsPolicy)
-        m_permissionsPolicy = makeUnique<PermissionsPolicy>(ownerElement(), securityOrigin().data());
```
Maybe instead, PermissionsPolicy could take a Document& and below logic would go in PermissionsPolicy constructor.

youennf commented on Source/WebCore/html/HTMLIFrameElement.cpp:
```diff
@@ -252,6 +254,21 @@ LazyLoadFrameObserver& HTMLIFrameElement::lazyLoadFrameObserver()
+    if (!remoteContentFrame)
+        return;
+
+    WebCore::OwnerPermissionsPolicyData policy;
```
```suggestion
    OwnerPermissionsPolicyData policy;
```
Also, why not initializing the members directly in constructor?

youennf commented on Source/WebCore/html/HTMLIFrameElement.h:
```diff
@@ -44,7 +44,9 @@ class HTMLIFrameElement final : public HTMLFrameElementBase {
 
-    PermissionsPolicy::PolicyDirective permissionsPolicyDirective() const;
+    WEBCORE_EXPORT PermissionsPolicy::PolicyDirective permissionsPolicyDirective() const;
+    void resetPermissionsPolicyDirective();
```
Can we make it private?

szewai replied:
This has been removed.

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -280,6 +280,22 @@ static Ref<SecurityOrigin> declaredOrigin(const HTMLIFrameElement& iframe)
     return iframe.document().securityOrigin();
 }
 
+OwnerPermissionsPolicyData::OwnerPermissionsPolicyData()
```
```suggestion
OwnerPermissionsPolicyData::OwnerPermissionsPolicyData()
    documentInheritedPolicyValue(0)
```
?

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -280,6 +280,22 @@ static Ref<SecurityOrigin> declaredOrigin(const HTMLIFrameElement& iframe)
+        return std::optional<HashSet<SecurityOriginData>> { origins };
+    }, [&] (const auto&) {
+        return std::optional<HashSet<SecurityOriginData>> { };
+    }), m_origins);
```
This copies everything. Maybe we should call it copyValue() or copyOrigins() instead.

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -413,29 +451,27 @@ bool PermissionsPolicy::inheritedPolicyValueForFeature(Feature feature) const
-#endif
-    m_inheritedPolicy.set(index(Feature::PrivateToken), computeInheritedPolicyValueInContainer(Feature::PrivateToken, frame, origin));
+    for (size_t index = 0; index < numberOfFeatures; ++index)
+        m_inheritedPolicy.set(index, computeInheritedPolicyValueInContainer(feature(index), frame, origin));
```
I am not sure I like BitSet, it removes enum typing for instance and makes the code less readable.
We could have a simple struct of booleans that would be as compact as a bitset.
And utility routines like `forEach(const Function<void(Feature)>&)` might make it convenient enough so that adding a new Feature will be only a few lines of code.

youennf commented on Source/WebCore/html/PermissionsPolicy.h:
```diff
@@ -90,11 +105,19 @@ class PermissionsPolicy {
+
 private:
+
+    PermissionsPolicy(WTF::BitSet<numberOfFeatures> inheritedPolicy)
```
explicit probably

youennf commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1350,6 +1350,11 @@ OptionSet<AdvancedPrivacyProtections> LocalFrame::advancedPrivacyProtections() c
     return { };
 }
 
+void LocalFrame::setOwnerPermissionsPolicy(const OwnerPermissionsPolicyData& policy)
```
Can we pass a &&, if it comes straight from IPC, it might save us a copy.

youennf commented on Source/WebCore/html/PermissionsPolicy.cpp:
```diff
@@ -449,4 +485,23 @@ bool PermissionsPolicy::isFeatureEnabled(Feature feature, const Document& docume
+{
+    HashMap<uint8_t, std::optional<HashSet<SecurityOriginData>>> result;
+    for (auto& [feature, allowlist] : policyDirective)
+        result.set(index(feature), allowlist.value());
```
We lose typing, I am not sure why.
Should probably be add as well, no need for set.

youennf commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -428,14 +428,20 @@ void WebFrameProxy::prepareForProvisionalLoadInProcess(WebProcessProxy& process,
     m_provisionalFrame = nullptr;
     m_provisionalFrame = makeUnique<ProvisionalFrameProxy>(*this, group.ensureProcessForSite(navigationSite, process, page->preferences()));
-    page->websiteDataStore().protectedNetworkProcess()->addAllowedFirstPartyForCookies(process, mainFrameDomain, LoadedWebArchive::No, WTFMove(completionHandler));
+    protectedProcess()->sendWithAsyncReply(Messages::WebPage::GetFrameOwnerPermissionsPolicy(m_frameID), [weakThis = WeakPtr { *this }, page = m_page, process = Ref { process }, frameID = m_frameID, mainFrameDomain = WTFMove(mainFrameDomain), completionHandler = WTFMove(completionHandler)](auto policy) mutable {
```
weakThis is not used.

youennf commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -428,14 +428,20 @@ void WebFrameProxy::prepareForProvisionalLoadInProcess(WebProcessProxy& process,
     m_provisionalFrame = makeUnique<ProvisionalFrameProxy>(*this, group.ensureProcessForSite(navigationSite, process, page->preferences()));
-    page->websiteDataStore().protectedNetworkProcess()->addAllowedFirstPartyForCookies(process, mainFrameDomain, LoadedWebArchive::No, WTFMove(completionHandler));
+    protectedProcess()->sendWithAsyncReply(Messages::WebPage::GetFrameOwnerPermissionsPolicy(m_frameID), [weakThis = WeakPtr { *this }, page = m_page, process = Ref { process }, frameID = m_frameID, mainFrameDomain = WTFMove(mainFrameDomain), completionHandler = WTFMove(completionHandler)](auto policy) mutable {
+        if (!page)
```
We probably want to take a Ref of the page here.

szewai replied:
This has been removed.

youennf commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -428,14 +428,20 @@ void WebFrameProxy::prepareForProvisionalLoadInProcess(WebProcessProxy& process,
+
+        process->send(Messages::WebPage::SetFrameOwnerPermissionsPolicy(frameID, policy), page->webPageIDInProcess(process));
+        page->websiteDataStore().protectedNetworkProcess()->addAllowedFirstPartyForCookies(process, mainFrameDomain, LoadedWebArchive::No, WTFMove(completionHandler));
+    }, page->webPageIDInMainFrameProcess());
```
I am not sure we need to send permission policy at the time we prepare the provisional load.
We only need to send it when starting to process the response data.
Maybe we should send it at the time the UIProcess receives the response policy check?

If the policies are not too big, maybe we could send them as part of the decide for policy action check, like we do for other iframe states?

szewai replied:
The latest patch includes the owner policy in navigation request instead; by doing that we don't need independent messages to set and update owner's permissions policy in subframe processes.

youennf commented on Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
```diff
@@ -65,6 +65,12 @@ void WebRemoteFrameClient::sizeDidChange(IntSize size)
 
+void WebRemoteFrameClient::ownerPermissionsPolicyDidChange(const OwnerPermissionsPolicyData& ownerPolicy)
+{
+    if (auto* page = m_frame->page())
```
```suggestion
    if (RefPtr page = m_frame->page())
```

youennf commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9752,6 +9753,44 @@ void WebPage::hasActiveNowPlayingSessionChanged(bool hasActiveNowPlayingSession)
 
+void WebPage::getFrameOwnerPermissionsPolicy(WebCore::FrameIdentifier frameID, CompletionHandler<void(const WebCore::OwnerPermissionsPolicyData&)>&& completionHandler)
+{
+    WebCore::OwnerPermissionsPolicyData policy;
```
I would tend to move it where we actually set members. Maybe we could use a constructor.
And use `completionHandler({ })` for the error cases.

youennf commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9752,6 +9753,44 @@ void WebPage::hasActiveNowPlayingSessionChanged(bool hasActiveNowPlayingSession)
+        return;
+
+    if (RefPtr coreFrame = frame->coreLocalFrame()) {
+        coreFrame->setOwnerPermissionsPolicy(policy);
```
We can probably use a OwnerPermissionsPolicyData&& and move it here or below

youennf commented on Source/WebCore/page/Frame.cpp:
```diff
@@ -257,4 +258,27 @@ void Frame::setOwnerElement(HTMLFrameOwnerElement* element)
+    if (!owner)
+        return std::nullopt;
+
+    auto documentOrigin = WebCore::SecurityOriginData { owner->document().securityOrigin().data() };
```
```suggestion
    auto documentOrigin = owner->document().securityOrigin().data();
```
No need for WebCore:: also

youennf commented on Source/WebCore/page/Frame.cpp:
```diff
@@ -257,4 +258,27 @@ void Frame::setOwnerElement(HTMLFrameOwnerElement* element)
+void Frame::setOwnerPermissionsPolicy(OwnerPermissionsPolicyData&& ownerPermissionsPolicy)
+{
+    // If the frame is in the same process as parent, it should compute policy from owner element directly.
+    if (dynamicDowncast<RemoteFrame>(tree().parent()))
```
Is this check needed? Is it hurting to set the owner policy for local frame as well when getting the info from UIProcess?

szewai replied:
This check aims to avoid any behavior change in non-site isolation case; maybe this is not needed. I will run tests to verify.

youennf commented on Source/WebCore/page/Frame.cpp:
```diff
@@ -257,4 +258,27 @@ void Frame::setOwnerElement(HTMLFrameOwnerElement* element)
+        m_ownerPermisssionsPolicy = WTFMove(ownerPermissionsPolicy);
+}
+
+std::optional<OwnerPermissionsPolicyData> Frame::ownerPermissionsPolicy()
```
Could we return a const std::optional<>?
Can this method be const?

youennf commented on Source/WebCore/page/Frame.cpp:
```diff
@@ -257,4 +258,27 @@ void Frame::setOwnerElement(HTMLFrameOwnerElement* element)
+    auto documentPolicy = owner->document().permissionsPolicy();
+    RefPtr iframe = dynamicDowncast<HTMLIFrameElement>(owner);
+    auto containerPolicy = iframe ? iframe->permissionsPolicyDirective() : WebCore::PermissionsPolicy::PolicyDirective { };
+    return OwnerPermissionsPolicyData { WTFMove(documentOrigin), WTFMove(documentPolicy), WTFMove(containerPolicy) };
```
Could we set m_ownerPermisssionsPolicy to this value?

szewai replied:
We probably should not cache m_ownerPermisssionsPolicy because frame that resides in the same process as its parent should be able to get the latest policy value by computation. `m_ownerPermisssionsPolicy` should only be used for frame that cannot access its parent. 
I will rename the parameters to make it more clear.

youennf commented on Source/WebCore/page/Frame.cpp:
```diff
@@ -257,4 +258,27 @@ void Frame::setOwnerElement(HTMLFrameOwnerElement* element)
+    auto documentOrigin = WebCore::SecurityOriginData { owner->document().securityOrigin().data() };
+    auto documentPolicy = owner->document().permissionsPolicy();
+    RefPtr iframe = dynamicDowncast<HTMLIFrameElement>(owner);
+    auto containerPolicy = iframe ? iframe->permissionsPolicyDirective() : WebCore::PermissionsPolicy::PolicyDirective { };
```
```suggestion
    auto containerPolicy = iframe ? iframe->permissionsPolicyDirective() : PermissionsPolicy::PolicyDirective { };
```

youennf commented on Source/WebCore/page/Frame.h:
```diff
@@ -137,6 +141,7 @@ class Frame : public ThreadSafeRefCounted<Frame, WTF::DestructionThread::Main>,
     WeakPtr<Frame> m_opener;
     WeakHashSet<Frame> m_openedFrames;
     mutable UniqueRef<HistoryController> m_history;
+    std::optional<OwnerPermissionsPolicyData> m_ownerPermisssionsPolicy;
```
Maybe make it mutable and make ownerPermissionsPolicy() const?

youennf commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -2073,6 +2073,9 @@ void WebPage::loadRequest(LoadParameters&& loadParameters)
             localMainFrame->loader().forceSandboxFlags(loadParameters.effectiveSandboxFlags);
     }
 
+    if (auto onwerPermissionsPolicy = std::exchange(loadParameters.ownerPermissionsPolicy, std::nullopt))
```
I tend to prefer:
```suggestion
    if (auto onwerPermissionsPolicy = std::exchange(loadParameters.ownerPermissionsPolicy, { }))
```

szewai reviewed (commented):
Seems this change affects non-site-isolation cases, I will look into test failures

szewai committed:
[Site Isolation] Make permissions policy work when site isolation is enabled
https://bugs.webkit.org/show_bug.cgi?id=276203
rdar://131080252

Reviewed by Youenn Fablet.

When site isolation is enabled, subframe can be in a different process than its parent. However, for document of the
subframe to compute permissions policy, it needs policy information of its owner element (parent frame). To make
permissions policy work with site isolation, we include owner's policy in navigation data, which is sent from web
process to UI process on deciding policy for navigation. UI process can then forward the owner's policy to correct frame
process by including it in LoadParameters.

Tests: http/tests/site-isolation/permissions-policy.html
       http/tests/site-isolation/permissions-policy-nested.html

* LayoutTests/http/tests/site-isolation/permissions-policy-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/permissions-policy-nested-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/permissions-policy-nested.html: Added.
* LayoutTests/http/tests/site-isolation/permissions-policy.html: Added.
* LayoutTests/http/tests/site-isolation/resources/permissions-policy-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/resources/permissions-policy-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/permissions-policy-nested-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/simple.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ownerPermissionsPolicy const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::attributeChanged):
(WebCore::HTMLIFrameElement::permissionsPolicyDirective const): Deleted.
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::PermissionsPolicy::computeInheritedPolicyValueInContainer const):
(WebCore::PermissionsPolicy::PermissionsPolicy):
* Source/WebCore/html/PermissionsPolicy.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::setOwnerPermissionsPolicy):
(WebCore::Frame::ownerPermissionsPolicy const):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::ownerPermissionsPolicy const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJSHistoryAPI):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/280942@main
